File: web/src/types/BlockRule.ts
import type BlockParserState from "./BlockParserState";
import type MarkdownNode from "./MarkdownNode";

// TODO: Try to arrange rules so that they don't have to refer to other rules
// e.g. checking for code in lists
export default interface BlockRule {
	/**
	 * The name for this rule, which must also be used for nodes created by this
	 * rule.
	 */
	name: string;
	/**
	 * The names of rules that this node closes.
	 */
	// TODO: closes: string[];
	/**
	 * Tests whether a node should start e.g. a block quote should start when we
	 * find a '>'.
	 * @param state
	 * @param parent
	 * @returns
	 */
	testStart: (state: BlockParserState, parent: MarkdownNode) => boolean;
	/**
	 * Creates a node for this rule.
	 * @param state
	 * @param parent
	 * @returns
	 */
	// TODO: createNode: (state: ParserState, parent: MarkdownNode) => void;
	/**
	 * Tests whether a node should continue after being started e.g. a block
	 * quote should continue if we find a '>'.
	 * @param state
	 * @param parent
	 * @returns
	 */
	testContinue: (state: BlockParserState, parent: MarkdownNode) => boolean;
	/**
	 * Does any cleanup for this rule's node when it is closed.
	 * @param state
	 * @param parent
	 * @returns
	 */
	closeNode?: (state: BlockParserState, parent: MarkdownNode) => void;
}

File: zig/src/types/BlockRule.zig
const std = @import("std");

pub const BlockRule = struct {
    const Self = @This();

    const BlockParserState = @import("BlockParserState.zig").BlockParserState;
    const MarkdownNode = @import("MarkdownNode.zig").MarkdownNode;

    name: []const u8,

    testStart: *const fn (state: *BlockParserState, parent: *MarkdownNode, end_of_line: usize) bool,

    testContinue: *const fn (state: *BlockParserState, parent: *MarkdownNode) bool,

    closeNode: ?*const fn (state: *BlockParserState, parent: *MarkdownNode) void = null,
};